synapse/config/workers.py

```diff
@@ -296,6 +306,43 @@ def read_config(self, config, **kwargs):
             self.worker_name is None and background_tasks_instance == "master"
         ) or self.worker_name == background_tasks_instance
 
+        # update_user_directory_on controls which worker is responsible for updating the user directory.
+        # None means that the main process should handle this, so does the absence of the config key.
+        #
+        # Note that due to backwards compatibility, we must also test for update_user_directory, and apply it if
+        # update_user_directory_on is not defined at the same time.
+        self.update_user_directory: bool
+
+        uud_result: bool
+        is_master_process = self.worker_name is None
+
+        if "update_user_directory_on" in config:
+            update_user_directory_on: Optional[str] = config["update_user_directory_on"]
+
+            if update_user_directory_on is None:
+                uud_result = is_master_process
+            else:
+                uud_result = self.worker_name == update_user_directory_on
+
+        else:
+            if "update_user_directory" in config:
+                # Backwards compatibility case
+                logger.warning(USER_UPDATE_DIRECTORY_DEPRECATION_WARNING)
+
+                update_user_directory: bool = config["update_user_directory"]
+
+                if update_user_directory:
+                    # Main process should update
+                    uud_result = is_master_process
+                else:
+                    # user directory worker should update
+                    uud_result = self.worker_app == "synapse.app.user_dir"
+            else:
+                # Both values are undefined, assume None for update_user_directory_on
+                uud_result = is_master_process
```

Comment:
this seems a very verbose, and not very clear, way of writing this logic.

```suggestion
        if "update_user_directory" in config:
            # Backwards compatibility case
            logger.warning(USER_UPDATE_DIRECTORY_DEPRECATION_WARNING)

            if config["update_user_directory"]:
                # Main process should update
                update_user_directory_on = "master"
            else:
                # user directory worker should update
                update_user_directory_on = "synapse.app.user_dir"
        else:
            update_user_directory_on = config.get("update_user_directory_on") or "master"
        
        self.update_user_directory = (
            is_master_process and update_user_directory_on == "master"
        ) or self.worker_name == update_user_directory_on
```

Reply:
This'd make the above `self.update_user_directory: bool` assertion also redundant.